### ISSUE TITLE
Bug in function "bbox_safe_rotate"

### ISSUE THREAD

**Issue by Yun-960** [albumentations-team/albumentations#961]
![image](https://user-images.githubusercontent.com/52908171/125404229-fce95500-e3e8-11eb-8011-aa634c458d52.png)

**Comment by BloodAxe** [comment]
Can you please elaborate on what the bug actually about and case to reproduce it?

**Comment by nordmannnick** [comment]
I have issues if I augment an image with settings of:

`transform = A.Compose( [A.SafeRotate(limit=45, p=1, border_mode=cv2.BORDER_CONSTANT), box_params=A.BboxParams(format='albumentations', label_fields=['gt_labels']) )`

I have spent quite a while tracking down the behaviour and hopefully it's an easy fix.
I am using 1.0.2 installed via PyCharm's package manager (a pip backend). The bug is also present in 1.0.1, 1.0.0 and 0.5.2 (I gave up testing further back after those).
I can successfully use HorizontalFlip, VerticalFlip and RandomSizedBBoxSafeCrop with no issues using the same surrounding code.

I have edited augmentations/geometric/functional.py as per Yun's screenshot and it does not seem to solve the issue for me.
Let me know if you need more information.

**Comment by BloodAxe** [comment]
Can you attach a stack trace, input bboxes and image shape please?

Вт, 13 июля 2021 г. в 21:13, nordmannnick ***@***.***>:

**Comment by Yun-960** [comment]
> Can you please elaborate on what the bug actually about and case to reproduce it?

It seems that the rescaling of the bbox is necessary. Please refer to the following code.

`random.seed(1)

image = np.ones((200,200,3), dtype=np.uint8) * 255
bboxes = np.array([[195  ,   195 , 199  ,  199  ]],
      dtype=np.float32)
class_labels = np.zeros((bboxes.shape[0],), dtype=np.int32)
catted_boxes = np.concatenate((bboxes, class_labels.reshape(-1,1)), axis=1)

transform = A.Compose([
    A.SafeRotate(limit=10, border_mode=0,value=0 ,p=1),
], bbox_params=A.BboxParams(format='pascal_voc', min_visibility=0.0), p=1)

transformed = transform(image=image, bboxes=catted_boxes)
transformed_image = transformed['image']

bboxes = transformed['bboxes']
assert len(bboxes) == catted_boxes.shape[0]`

![image](https://user-images.githubusercontent.com/52908171/125546800-84789e47-a13d-46b7-a34f-1395b1009fef.png)
![image](https://user-images.githubusercontent.com/52908171/125546848-8bae3ab8-ee5c-4362-97dc-13dac199dfb6.png)
![image](https://user-images.githubusercontent.com/52908171/125546890-3894ff51-c884-4267-b8d6-f2b6d4ca3d6c.png)

**Comment by Anjum48** [comment]
I am also seeing this issue using Yun's code example. Sometimes the box is successfully rotated (usually for small angles), but for the majority of the time, an empty array is returned for the bbox.

**Comment by nordmannnick** [comment]
What further information do you need to progress on this bug? My experience of using the function is that it just flat out does not work as expected (the bounding box does not rotate with the object in the image). Yun's code seems to be a minimum working example to replicate the issue and the behaviour seems to predate 1.0.0.

I can only presume the unit test for the function isn't functioning as you expect.

**Comment by vriez** [comment]
The `SafeRotate` transform is problematic also for the transformed bounding boxes.

The workaround I used to get past this behavior was to pad the images at all edges.

I debugged it by adding `keypoints` to each corner and plotting the transformed image with the keypoints on it.


It looked something like:
![16343644648013](https://user-images.githubusercontent.com/5475058/159225291-b82fe614-706c-406d-90a6-5eb804aaefaf.png)
